File: tp2/MMU.java
package bcc266TP2.toy;

//MMU -> memory management unit
//UGM -> unidade de gerenciamento de memoria
public class MMU {
	
	
	static BlocoMemoria buscarNasMemorias(Endereco e, RAM ram, BlocoMemoria[] cache1, BlocoMemoria[] cache2){
		
		//mapeamento direto a memoria
		int posicaoCache1 = e.endBloco%cache1.length;
		int posicaoCache2 = e.endBloco%cache2.length;
		int custo = 0;
		
		if(cache1[posicaoCache1].endBloco==e.endBloco){
			custo =10;
			cache1[posicaoCache1].custo = custo;
			//indica onde foi achado 
			// 1 significa cache 1; 2 significa cache 2; e 3 significa RAM
			cache1[posicaoCache1].cacheHit = 1;
			return cache1[posicaoCache1];
		}else{
			
			if(cache2[posicaoCache2].endBloco==e.endBloco){	
				custo =110; // inclui custo da cache 1
				cache2[posicaoCache2].cacheHit = 2;
				return movCache2Cache1(posicaoCache1, posicaoCache2, cache1, cache2, custo);
			}else
			{
				//buscar na ram e levar para cache2 e depois levar para cache1
				custo =1110;				
				return movRamCache2(posicaoCache1, posicaoCache2, cache1, cache2, ram, e, custo);
				
			}			
		}	
		
	}
	
	static BlocoMemoria movCache2Cache1(int posicaoCache1, int posicaoCache2, BlocoMemoria[] cache1, BlocoMemoria[] cache2, int custo){
		BlocoMemoria aux = cache1[posicaoCache1];	
		cache1[posicaoCache1] = cache2[posicaoCache2];
		cache2[posicaoCache2] = aux;
		
		cache1[posicaoCache1].custo = custo;
		
		return cache1[posicaoCache1]; 
	}	
	
	static BlocoMemoria movRamCache2(int posicaoCache1, int posicaoCache2, BlocoMemoria[] cache1, BlocoMemoria[] cache2, RAM ram, Endereco e, int custo){
		if(!cache2[posicaoCache2].atualizado){
			cache2[posicaoCache2] = ram.getDado(e.endBloco);
			cache2[posicaoCache2].cacheHit = 3;
			return movCache2Cache1(posicaoCache1, posicaoCache2, cache1, cache2, custo);
			
		}else{
			cache2[posicaoCache2].atualizado = false;
			cache2[posicaoCache2].cacheHit = 0;
			cache2[posicaoCache2].custo = 0;
			ram.setDado(cache2[posicaoCache2].endBloco, cache2[posicaoCache2]);
					
			cache2[posicaoCache2] = ram.getDado(e.endBloco);
			cache2[posicaoCache2].cacheHit = 3;
			
			return movCache2Cache1(posicaoCache1, posicaoCache2, cache1, cache2, custo);
		}		 
	}

}

File: tp2/MMU.c
#include "cpu.h"


int verificarMaisVelho(int nCache, CPU *cpu)
{
    int maisVelho = 0;

    switch (nCache)
    {
    case 1:
        for (int i = 0; i < TAM_C1; i++)
        {
            if ( cpu->cache1.vetorDeBlocos[i].ultimaVezAcessado < cpu->cache1.vetorDeBlocos[maisVelho].ultimaVezAcessado 
             ||  cpu->cache1.vetorDeBlocos[i].endBloco == -1 )
            {
                maisVelho = i;
            }
        }
        break;
    case 2:
        for (int i = 0; i < TAM_C2; i++)
        {
            if ( cpu->cache2.vetorDeBlocos[i].ultimaVezAcessado < cpu->cache2.vetorDeBlocos[maisVelho].ultimaVezAcessado 
             ||  cpu->cache2.vetorDeBlocos[i].endBloco == -1 )
            {
                maisVelho = i;
            }
        }
        break;
    case 3:
        for (int i = 0; i < TAM_C3; i++)
        {
            if ( cpu->cache3.vetorDeBlocos[i].ultimaVezAcessado < cpu->cache3.vetorDeBlocos[maisVelho].ultimaVezAcessado 
             ||  cpu->cache3.vetorDeBlocos[i].endBloco == -1 )
            {
                maisVelho = i;
            }
        }
        break;
    default:
        break;
    }

    return maisVelho;
}

void trocarBloco(Memoria* origem, Memoria* destino, int pos, int maisVelho)
{
    BlocoDePalavras aux = origem->vetorDeBlocos[pos];

    origem->vetorDeBlocos[pos] = destino->vetorDeBlocos[maisVelho];
    destino->vetorDeBlocos[maisVelho] = aux;
    
    return;
}

BlocoDePalavras* movCache2Cache1(CPU *cpu, int endBloco, int maisVelho)
{
    trocarBloco(&cpu->cache2, &cpu->cache1, endBloco, maisVelho);

    cpu->cache1.vetorDeBlocos[maisVelho].ultimaVezAcessado = ++cpu->tempoAtual;

    return &cpu->cache1.vetorDeBlocos[maisVelho];
}

BlocoDePalavras* movCache3Cache2(CPU *cpu, int endBloco, int maisVelho)
{
    trocarBloco(&cpu->cache3, &cpu->cache2, endBloco, maisVelho);

    return &cpu->cache2.vetorDeBlocos[maisVelho];
}

BlocoDePalavras* movRAMCache3(CPU *cpu, int endBloco, int maisVelho)
{
    // trocarBloco(&cpu->RAM, &cpu->cache3, endBloco, maisVelho);

    if (cpu->cache3.vetorDeBlocos[maisVelho].endBloco == -1)
    {
        cpu->cache3.vetorDeBlocos[maisVelho] = cpu->RAM.vetorDeBlocos[endBloco];;
    }else if (cpu->cache3.vetorDeBlocos[maisVelho].atualizado)
    {
        //*Atualiza a RAM
        cpu->cache3.vetorDeBlocos[maisVelho].atualizado = false;
        cpu->RAM.vetorDeBlocos[cpu->cache3.vetorDeBlocos[maisVelho].endBloco] = cpu->cache3.vetorDeBlocos[maisVelho];
        cpu->cache3.vetorDeBlocos[maisVelho] = cpu->RAM.vetorDeBlocos[endBloco];
    }else
    {
        cpu->cache3.vetorDeBlocos[maisVelho] = cpu->RAM.vetorDeBlocos[endBloco];
    }


    return &cpu->cache3.vetorDeBlocos[maisVelho];
}

BlocoDePalavras* buscarNasMemorias(CPU *cpu, Endereco end)
{
    for (int i = 0; i < TAM_C1; i++)
    {
        if (cpu->cache1.vetorDeBlocos[i].endBloco == end.endBloco)
        {
            cpu->cache1.cacheHit++;
            cpu->custo += TAM_C1;
            
            return &cpu->cache1.vetorDeBlocos[i];
        }
    }
    cpu->cache1.cacheMiss++;
    cpu->custo += TAM_C1;

    for (int i = 0; i < TAM_C2; i++)
    {
        if (cpu->cache2.vetorDeBlocos[i].endBloco == end.endBloco)
        {
            cpu->cache2.cacheHit++;
            cpu->custo += TAM_C2;
            int maisVelho = verificarMaisVelho(1, cpu);
            return movCache2Cache1(cpu, i, maisVelho);
        }
    }
    cpu->cache2.cacheMiss++;
    cpu->custo += TAM_C2;

    for (int i = 0; i < TAM_C3; i++)
    {
        if (cpu->cache3.vetorDeBlocos[i].endBloco == end.endBloco)
        {
            cpu->cache3.cacheHit++;
            cpu->custo += TAM_C3;

            int maisVelho = verificarMaisVelho(2, cpu);

            movCache3Cache2(cpu, i, maisVelho);

            maisVelho = verificarMaisVelho(1, cpu);

            return movCache2Cache1(cpu, i, maisVelho);
        }
    }
    cpu->cache3.cacheMiss++;
    cpu->custo += TAM_C3;
    cpu->custo += TAM_RAM;

    int maisVelho = verificarMaisVelho(3, cpu);
    movRAMCache3(cpu, end.endBloco, maisVelho);

    maisVelho = verificarMaisVelho(2, cpu);
    movCache3Cache2(cpu, end.endBloco, maisVelho);

    maisVelho = verificarMaisVelho(1, cpu);
    return movCache2Cache1(cpu, end.endBloco, maisVelho);
}
